Source/NameServer/AddressBook.hpp

```diff
@@ -0,0 +1,166 @@
+    unsigned FindDevice(std::string &TaskName, std::string &Name, const Record_t* &DRec);
+    unsigned FindDevice(std::string &TaskName, SymAddr_t Address, const Record_t* &DRec);
+
+    // Find all devices supervised by X
+    unsigned FindBySuper(std::string &TaskName, SymAddr_t Supervisor, const RecordVect_t* &Records);
+
+    unsigned FindByType(std::string &TaskName, std::string Type, const RecordVect_t* &Records);
+    unsigned FindByAttribute(std::string &TaskName, std::string Attribute, const RecordVect_t* &Records);
+
+    unsigned FindByInMsg(std::string &TaskName, std::string Msg, const RecordVect_t* &Records);            //TODO
+    unsigned FindByOuMsg(std::string &TaskName, std::string Msg, const RecordVect_t* &Records);            //TODO
+
+    // device getters. Note that a vector<Record_t> is different from a RecordVect_t,
+    // which is a vector<const Record_t*>
+    unsigned GetDevices(std::string &TaskName, const std::vector<Record_t>* &Records);
+    unsigned GetExternals(std::string &TaskName, const std::vector<Record_t>* &Records);
+    unsigned GetSupervisors(std::string &TaskName, const std::vector<Record_t>* &Records);
+    unsigned GetExtCon(std::string &TaskName, const RecordVect_t* &Records);
+
+    // Device Count getters
+    long GetDeviceCount(std::string &TaskName);
+    long GetLoadedDeviceCount(std::string &TaskName);
+    long GetExternalCount(std::string &TaskName);
+    long GetLoadedExternalCount(std::string &TaskName);
+    int GetSupervisorCount(std::string &TaskName);
+
+    // Integrity
+    unsigned IntegCheck(void);                  // Integrity check of ALL tasks
+    unsigned IntegCheck(std::vector<unsigned> &TaskIntegs);
+    unsigned IntegTask(std::string &TaskName, bool Verbose = false, FILE * = stdout);  // Ingerity Check a Task by name
+  //unsigned IntegTask(const TaskRecord_t *TRec, bool Verbose = false, FILE * = stdout);     // Ingerity Check a Task by pointer
+
+    // Debugging
+    void Dump(std::string &Name);
+    void Dump(FILE * = stdout, std::string Name = "");
+
+    static const long     ERR_BAD_COUNT = -1;
+  
+    static const unsigned SUCCESS = 0;
+    static const unsigned ERR_NONFATAL = 0;
+    static const unsigned ERR_INVALID_TASK = 1;
+    static const unsigned ERR_INVALID_DEVTYPE = 2;
+    static const unsigned ERR_INVALID_DEVICE = 3;
+    static const unsigned ERR_INVALID_MESSAGE_TYPE = 4;
+    static const unsigned ERR_INVALID_ATTRIBUTE = 5;
+    static const unsigned ERR_DEVICE_DATA_MISMATCH = 6;
+    static const unsigned ERR_TASKNAME_USED = 7;
+    static const unsigned ERR_DEVICENAME_USED = 8;
+    static const unsigned ERR_DEVICE_ADDR_USED = 9;
+    static const unsigned ERR_TASK_NOT_FOUND = 10;
+    static const unsigned ERR_DEVICE_NOT_FOUND = 11;
+    static const unsigned ERR_INVALID_MAP = 12;
+    static const unsigned ERR_INVALID_SUPERVISOR = 13;
+    static const unsigned ERR_INVALID_STATE = 14;
+
+private:
+    TaskMap_t TaskMap;
+    int TaskCount;
```

Comment:
Why do you need to maintain/update this variable? Surely you can just do `TaskMap_t.size()`.

Reply:
Graeme? Does this have an important use? Are you just trying to minimise the calls to the size() function?

Reply:
I think I was trying to minimise the calls to size() - it is also a good metric to looks for "oopsies" to the data structure.